server/src/main/java/org/elasticsearch/gateway/LucenePersistedStateFactory.java

```diff
@@ -0,0 +1,523 @@
+                    }
+                    addIndexMetaData(indexMetaData);
+                } else {
+                    logger.trace("no action required for [{}]", indexMetaData.getIndex());
+                }
+                indexMetadataVersionByUUID.remove(indexMetaData.getIndexUUID());
+            }
+
+            for (final String removedIndexUUID : indexMetadataVersionByUUID.keySet()) {
+                logger.trace("removing metadata for [{}]", removedIndexUUID);
+                deleteIndexMetaData(removedIndexUUID);
+            }
+
+            // Flush, to try and expose a failure (e.g. out of disk space) before committing, because we can handle a failure here more
+            // gracefully than one that occurs during the commit process.
+            indexWriter.flush();
+        }
+
+        void persist(String nodeId, long currentTerm, long lastAcceptedVersion) throws IOException {
+            final Map<String, String> commitData = new HashMap<>(2);
+            commitData.put(CURRENT_TERM_KEY, Long.toString(currentTerm));
+            commitData.put(LAST_ACCEPTED_VERSION_KEY, Long.toString(lastAcceptedVersion));
+            commitData.put(NODE_VERSION_KEY, Integer.toString(Version.CURRENT.id));
+            commitData.put(NODE_ID_KEY, nodeId);
+            indexWriter.setLiveCommitData(commitData.entrySet());
+            indexWriter.commit();
+        }
+
+        @Override
+        public void close() throws IOException {
+            IOUtils.close(indexWriter, directory);
+        }
+
+        private void deleteGlobalMetaData() throws IOException {
+            indexWriter.deleteDocuments(new Term(TYPE_FIELD_NAME, GLOBAL_TYPE_NAME));
+        }
+
+        private void addGlobalMetaData(MetaData metaData) throws IOException {
+            indexWriter.addDocument(buildDocument(GLOBAL_TYPE_NAME, metaData));
+        }
+
+        private void deleteIndexMetaData(IndexMetaData indexMetaData) throws IOException {
+            deleteIndexMetaData(indexMetaData.getIndexUUID());
+        }
+
+        private void deleteIndexMetaData(String indexUUID) throws IOException {
+            indexWriter.deleteDocuments(new Term(INDEX_UUID_FIELD_NAME, indexUUID));
+        }
+
+        private void addIndexMetaData(IndexMetaData metaData) throws IOException {
+            final Document document = buildDocument(INDEX_TYPE_NAME, metaData);
+            assert metaData.getIndexUUID().equals(IndexMetaData.INDEX_UUID_NA_VALUE) == false;
+            document.add(new StringField(INDEX_UUID_FIELD_NAME, metaData.getIndexUUID(), Field.Store.NO));
+            indexWriter.addDocument(document);
+        }
+
+        private Document buildDocument(String typeName, ToXContent metaData) throws IOException {
+            final Document document = new Document();
+            document.add(new StringField(TYPE_FIELD_NAME, typeName, Field.Store.NO));
+
+            final ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
```

Comment:
The safety of this kind of sharing is unclear to me -- I can see that this works today but it's unclear it will still work safely in future versions. @jpountz does this sound safe (and future-proof) to you?

Reply:
Come to think of it, now that we're reusing the `document` across paths it's probably optimal to not reuse buffers here (because the CS will be much larger than the index metadata making us hold on to a large chunk of memory for longer than we have to?).
I'd still go for the non-copying `toByteArray` override here. That doesn't have any risk to it if we don't reuse the BAOS and saves a few larger allocations right?

Reply:
See 983825dcd62b4fa9e684052e3e362b9a10a2309c for an implementation based on `BigArrays` instead.